src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,141 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_FACTOR = 1.5;
+    private transient Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = new Object[DEFAULT_CAPACITY];
+    }
+
+    public ArrayList(int initCapacity) {
+        if (initCapacity <= 0) {
+            throw new IllegalArgumentException(
+                    "Can't create array with negative length");
+        }
+        elements = new Object[initCapacity];
+    }
+
     @Override
     public void add(T value) {
-
+        growIfNeeded();
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        addElementArrayListIndexOutOfBoundException(index);
+        growIfNeeded();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        growIfNeeded(list);
+        int listSize = list.size();
+        for (int i = 0; i < listSize; i++) {
+            elements[size + i] = list.get(i);
```

Comment:
Maybe you can use your add() method here. in this case you wouldn't need  growIfNeeded() overloading. But as soon I'm not sure I leave it fo mentor judging

Reply:
That is good point you can just call method add in loop body for adding elements of list provided